### PR TITLE
`General`: Allow instructors to define individual due dates for students in course exercises

### LINKED ISSUE #4196: RFC: Individual Extended Working Time for Exercises
# Is your feature request related to a problem?

In some lectures the points of multiple exercises performed during the semester is combined into the final grade (‘Portfolioprüfung’). As in those cases the exercise directly influences the grade, sick days the student can prove via a doctor’s notice have to be compensated by individually extending that student’s final submission date for the current exercise.

Extensions of working time is currently only possible for exams, but not not for regular exercises within Artemis.

# Describe the solution you'd like

## User Facing Changes
- As an instructor to the exercise
    - When accessing the ‘Participation’ page in the exercise management, I would like to see in the list of students that participate for which ones an individual due date has been set.
    - I would like to be able to add an individual due date to any participation of the exercise. The due date should only be allowed to be later than the regular exercise due date.
- As a tutor
    - After the regular due date not yet finished submissions should not show up when assessing other submissions of the same exercise.
- As a student
    - I would like to see my individual due date in the Artemis client.
    - I would like to be able to submit to the exercise until my individual due date.

## Changes to the data model
The individual working times for exams are stored in the `StudentExam`. Similarly for exercises, this can be done in `Participation` by adding a `ZonedDateTime individualDueDate` attribute.
As this is part of the `Participation`, the student is at least required to click on ‘Start Exercise’ before the regular due date to make sure such a `Participation` for the student exists. This is a reasonable and acceptable limitation.

## Effects that have to be considered
- The uniform due date for an exercise is assumed at many places throughout most of Artemis’s functionality.
- All exercises: Assessment on other submissions should be able to start before the last student has finished their exercise. The not yet finished participation should not yet be possible to be assessed by tutors.
- Programming exercises: When receiving build results, the test cases and feedbacks for tests marked with `AFTER_DUE_DATE` have to be filtered according to the individual due date.
- Programming exercises: The scheduling for the after due date tasks has to be updated:
   - The run tests after due date CI-runs have to be scheduled according to either the specified date, or after the individual due date depending on which one exists.
   - The repository locking mechanism after the due date has to exclude those repositories that belong to students that are still allowed to work on the exercise.
   - The repository locking mechanism has to lock the repository on the individual due date.
- Possible effects on the notification system? → I’ve not looked into it, but that is not a core functionality and necessary changes could be implemented in a follow-up PR.
- Individual due dates should only be possible for exercises that actually have a due date themself.
- Moving the due date for the exercise itself to a later point in time should remove all individual due dates that are before the general due date.
    - Invariant: ∀ i ∈ individual due dates: i = `null` ∨ i > exercise due date

### Server
Similarly to the existing `ExamDateService` a new `ExerciseDateService` should be introduced. Then all server-side checks if the due date has passed can be directed to a central location.
The check needs the student participation. If no individual due date is set for that participation, then the due date of the exercise is used. If the exercise is an exam exercise the due date check from the `ExamDateService` should be used.

Apply a filter to the submissions when tutors request one for assessment. Not yet finished ones should not be able to be sent to the client for assessment.

Scheduling changes (only relevant for programming exercises):
- For all students in the exercise for which no individual due date is set the regular scheduling as implemented now should still be used.
    - Spawning a lot of individual tasks that concurrently perform basically the same actions produces unnecessary load on the system, especially for very big courses with many students.
    - Instead, when locking repositories, and when updating scores a filter has to be applied to only perform those actions on submissions that do not have an individual due date (can be done in the query fetching them from the database).
- Only for students which have got an individual due date, individual tasks for locking the repository and updating the score are scheduled.
    - This should only affect very few students, no extraordinary load is to be expected.
    - The `ScheduleService` should be extended to also allow for scheduling based on the `Participation`.
    - A new small `ParticipationLifecycleService` to handle those scheduled tasks will be introduced.

### Client
- Where a due date is shown to the student, this should now be the individual due date if one exists. If not, the regular exercise due date should be shown.
    - A similar check as in the server that uses the exercise and participation to determine the due date should be introduced.
- The ability to choose an individual due date for a student as instructor has to be added to the participation list.
- `AFTER_DUE_DATE`-feedbacks are filtered server-side. Requesting a result with feedbacks therefore works as before without changes to the client.
- When tutors request the next submission to assess, a request is made to the server anyway. The server can filter the submissions without assessments accordingly to not return a non-finished one. ⇒ No change needed.

# Describe alternatives you've considered

Manually unlocking a single repository and allowing the student to push further commits would be possible. Artemis at the moment does not handle the exclusion from manual assessments. In programming exercises with tests marked with `AFTER_DUE_DATE`, those would be visible to the student while still working on the exercise.
⇒ Not a real alternative.

# Additional context

I’ve marked this PR as RFC, as I want to change a core idea/assumption in the exercise model. Therefore, I want to make sure that possible effects have been considered not only be myself to make sure I did not miss potentially important ones before I start implementing the feature.

WIP on branch `feature/exercises/individual-due-dates`: https://github.com/ls1intum/Artemis/tree/feature/exercises/individual-due-dates, PR #4416

### PR DESCRIPTION
**Pull request by b-fein**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
The actual interaction (i.e. actual code that performs the repo locks) with Bamboo/Gitlab has not changed.
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 5 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on our staging setup (Jenkins and Gitlab).

### Motivation and Context
Closes #4196 
In some lectures the points of multiple exercises performed during the semester is combined into the final grade (‘Portfolioprüfung’). As in those cases the exercise directly influences the grade, sick days the student can prove via a doctor’s notice have to be compensated by individually extending that student’s final submission date for the current exercise.

### Description
The implementation draft as outlined in #4196 has been followed.
This means
- individual due dates are only possible for exercises that have got a regular due date.
- individual due dates must always be after the regular due date of the exercise.
- quiz exercises cannot have individual due dates.
- exam exercises cannot have individual due dates.

### Steps for Testing

Prerequisites:
- An Instructor
- A Tutor
- At least two students, one for which an individual due date will be set (A), another where the regular due date applies (B).

#### Quiz Exercises/Exam Exercises
1. Make sure the column where the individual due dates can be set are not visible in the participants list of the exercise.

#### Modelling/Text/File-Upload Exercises
1. As the instructor create an exercise of one of the types mentioned above.
    - The exercise should have a due date.
2. Start the exercise as both students before the due date.
3. As instructor change the individual due date for student (A) on the participants list of the exercise.
4. Student (B) should be able to submit solutions to the exercise only until the regular due date.
5. Student (A) should be able to continue to work on the exercise.
6. After the regular due date but before the individual due date has passed, start the assessment process. You should only e able to assess the solution of (B), i.e. after you assessed one solution the message ‘No further submissions left for assessment.’ should be shown.
7. After the individual due date has passed, (A) should no longer be able to make further submissions. 
8. As a tutor, you should now be able to assess an additional submission.

#### Programming Exercises
1. The steps above should apply as well. There are a few things to look out for:
    - The submissions both via Git and the online editor should work as described until the regular/individual due date.
    - If the exercise contains tests marked as AFTER_DUE_DATE (can be configured on the ‘Grading’ page of the exercise), their feedback should only be visible for automatic results after the due date of *all* students has passed to avoid an unfair advantage for students with longer working times.
    - However, the score itself should be updated on the due date/on the build and test date even if a later individual due date exists.
    - On the result details dialogue (c.f. screenshot below) a short explanation why some feedback details are still missing should be shown.
2. If you add an individual due date that is in the future for student (B) after the regular due date has passed, their repository should now be unlocked again and they can submit further revisions.

#### Additional small tests
1. If you change the regular exercise due date to a due date after an existing individual due date, the individual due date should be removed.
2. If you remove the due date from an exercise, all individual due dates should be removed.

In general, you should check in places where the due date is shown to students if the individual date is shown correctly, or if the component still shows the regular due date of the exercise instead.

#### Code Review
This PR got quite big even though I tried to move some necessary refactoring into separate earlier PRs if possible. More than half of the changes (1777+, 400-) made in this PR are unit tests for changed functionality that sometimes needed some more additional tests for related existing edge-cases that had not been tested properly before.
- [x] Review 1
- [x] Review 2
    - [x] Client Code
    - [x] Server Code

#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
| File | Branch | Statement |
| --- | --- | --- |
| `ProgrammingExerciseScheduleService` | 74% | 86% |
| `ProgrammingExerciseGradingService` | 85% | 94% |
| `ProgrammingExerciseParticipation` | 80% | 87% |
| `ExerciseDateService` | 91% | 98% |
| `ParticipationLifecycleService` | 81% | 97% |
| `ParticipationService` | 76% | 94% |
| `SubmissionService` | 78% | 93% |
| `ProgrammingExerciseExportService` | 63% | 90% |
| `ScheduleService` | 100% | 100% |
| `ParticipationResource` | 67% | 93% |
| `exercise.utils.ts` | 89.9% | 89.7% |
| `participation.component.ts` | 82.9% | 85.7% |
| `course-exercises.component.ts` | 77.6% | 98.8% |

In cases where the branch coverage is lower than 80% at least all new code has (nearly) full coverage.

### Screenshots
New feature enables adding individual due dates for students on the ‘Participations’ overview page:
Screenshot_2021-11-29 Participation
Additional information on the result details dialogue about test case feedback that may still be hidden until the submission due date is over for all students (only shown for automatic results after the regular due date for programming exercises):
Screenshot_2021-12-02 Code Editor

### CODE REVIEW

**Review by dfuchss — DISMISSED** [review]
Tested yesterday with some teaching assistants. Seems to work :)

**Review by dfuchss — APPROVED** [review]
I've looked into all code.

* Java code looks very good to me.
* Client side code as well (but I'm not very used to the client side code)

Next week we'll test the functionality again as well. Afterwards, I'll comment on it again :)

File: src/main/java/de/tum/in/www1/artemis/service/SubmissionService.java [hunk 1]
```diff
@@ -204,14 +206,17 @@ public void checkSubmissionLockLimit(long courseId) {
...
         if (submissionsWithoutResult.isEmpty()) {
             return Optional.empty();
         }
-
-        submissionsWithoutResult = selectOnlySubmissionsBeforeDueDateOrAll(submissionsWithoutResult, exercise.getDueDate());
-
-        var submissionWithoutResult = submissionsWithoutResult.get(random.nextInt(submissionsWithoutResult.size()));
-        return Optional.of(submissionWithoutResult);
+        else {
+            final Random random = new Random();
```
[the remark this answers is not in the document]
Comment (b-fein) [reply src/main/java/de/tum/in/www1/artemis/service/SubmissionService.java]: Thanks, it definitely is a smell. I replaced it with `ThreadLocalRandom.current()` in 5117e0922353ddfa1a8d8d216522a7af857c058b.

**Review by dfuchss — DISMISSED** [review]
Re-approve

**Review by dfuchss — DISMISSED** [review]
Re-Approve Code

**Review by dfuchss — DISMISSED** [review]
Re-Approve Code

**Comment by LDAP** [comment]
Before starting the exercise the UI states "No due date" which is misleading:
image
Note: That could also be the behavior before the PR.

**Comment by dfuchss** [comment]
@b-fein you may have to adapt the max sizes again. I'm currently unable to build it :)

`Error: bundle initial exceeded maximum budget. Budget 2.00 MB was not met by 49.78 kB with a total of 2.05 MB.`

See https://github.com/kit-sdq/Artemis4Docker/runs/4470076692?check_suite_focus=true

**Comment by b-fein** [comment]
> @b-fein you may have to adapt the max sizes again. I'm currently unable to build it :)

@dfuchss It got increased to 2.2mb on `develop` in the meantime, it should work again.

**Comment by dfuchss** [comment]
Yes, it builds again

**Comment by dfuchss** [comment]
Currently the SAML2 Feature is not shown on the branch:

`ERROR TypeError: fe['display-ribbon-on-profiles'] is undefined`

Shown: 
image

Expected:
image

**Comment by dfuchss** [comment]
Don't know whether this is related to the current `develop` or to this branch

**Comment by b-fein** [comment]
@dfuchss It looks like the error might originate here: https://github.com/ls1intum/Artemis/blob/627aabbc6946c20589b0e2715b98f7649ac6e6b8/src/main/webapp/app/shared/layouts/profiles/profile.service.ts#L29

That is an area that has definitely not been touched by this PR.

It looks like the `display-ribbon-on-profiles` should be part of the response of the `/management/info` endpoint (on my system on the stable version is), but that is no longer the case.

**Review by Malyuk-A — DISMISSED** [review]
Tested in today's Testing Session on TS3. Worked as described (including very tricky edge cases).
We found minor inconsistencies with a few due date UIs (main student exercise list) but these will be covered soon. 👍

**Review by dfuchss — APPROVED** [review]
Re-Approve since Testing Session went well. (besides the minor table date issue :))

**Comment by b-fein** [comment]
@dfuchss @Malyuk-A @willergeorg As discussed in the testing session earlier, there were a few locations where the individual due dates were not shown correctly in the course overview. I updated all three to now show the individual due dates. 

#### Exercises are now put in the correct weekly category when an individual due date is set
Before the changes:
Screenshot 2021-12-09 at 20-05-13 Course details
After:
Screenshot 2021-12-09 at 20-05-50 Course details

#### Sidebar
When the participation has no individual due date:
Screenshot 2021-12-09 at 20-31-38 Course details
After setting an individual due date that is even after the due date of another exercise:
Screenshot 2021-12-09 at 21-23-58 Course details

#### Headline
It now shows if the individual due date is in the next twelve hours. Difficult to produce a useful screenshot.

Combined diff for the described updates: https://github.com/ls1intum/Artemis/compare/5913dd5%5E%5E%5E%5E%5E..5913dd5

**Review by Malyuk-A — COMMENTED** [review]
Thank you for the very fast fix of our nitpicks.
I approve of your changes, but there are still a couple of Codacy issues left 😅

**Comment by b-fein** [comment]
> Thank you for the very fast fix of our nitpicks. I approve of your changes, but there are still a couple of Codacy issues left :sweat_smile:

@Malyuk-A Five of the Codacy requests are missing Javadoc for unit tests. As almost all other tests do not have a Javadoc either, I don’t really see a point in adding them here.
The one about the long parameter list is just a case where a service needs a lot of other repositories/services, which causes a long constructor parameter list. I don’t think that’s an actual issue.
The one about the too complex method I cannot really solve, as the method is just an if-chain checking conditions. It can’t really be simplified, and splitting it up doesn’t help to make the code more understandable in that case, I think. The method Codacy complains about:
https://github.com/ls1intum/Artemis/blob/5b4c9caa91de15bc1b66ed04230086b3b68c39b0/src/main/java/de/tum/in/www1/artemis/service/scheduled/ProgrammingExerciseScheduleService.java#L146

**Review by willergeorg — DISMISSED** [review]
Tested during testing session. Worked good.

**Review by dfuchss — DISMISSED** [review]
Re-Approve Code Functionality

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseParticipationResource.java [hunk 2]
```diff
@@ -99,19 +95,33 @@ public ProgrammingExerciseParticipationResource(ProgrammingExerciseParticipation
...
 
         return result.map(ResponseEntity::ok).orElseGet(() -> ResponseEntity.ok(null));
     }
 
+    /**
+     * Removes sensitive information that students should not see (yet) from the given result.
+     * @param participation the result belongs to.
+     * @param result the sensitive information of which should be removed.
+     */
+    private void filterResult(final Participation participation, final Result result) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseParticipationResource.java:114]: I would rename this method, e.g. filterSensitiveInformationInResult

File: src/main/resources/config/liquibase/master.xml [hunk 3]
```diff
@@ -145,8 +145,9 @@
     <include file="classpath:config/liquibase/changelog/20211029230000_changelog.xml" relativeToChangelogFile="false"/>
     <include file="classpath:config/liquibase/changelog/20211101080828_changelog.xml" relativeToChangelogFile="false"/>
     <include file="classpath:config/liquibase/changelog/20211107212500_changelog.xml" relativeToChangelogFile="false"/>
-    <include file="classpath:config/liquibase/changelog/20211126224000_changelog.xml" relativeToChangelogFile="false"/>
+    <include file="classpath:config/liquibase/changelog/20211111111111_changelog.xml" relativeToChangelogFile="false"/>
     <include file="classpath:config/liquibase/changelog/20211115200300_changelog.xml" relativeToChangelogFile="false"/>
+    <include file="classpath:config/liquibase/changelog/20211126224000_changelog.xml" relativeToChangelogFile="false"/>
```
Comment (krusche) [inline src/main/resources/config/liquibase/master.xml:150]: Can you add your changeling to the end and preserve the order?

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseParticipationResource.java (hunk 2 above)
Comment (b-fein) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseParticipationResource.java]: :+1: Done in 207454ec9c77b6eefb1fd83a04884f7acd42f4de.

**Review by dfuchss — APPROVED** [review]
Re-Approve

**Review by dfuchss — APPROVED** [review]
Re-Approve

**Review by dfuchss — DISMISSED** [review]
Changes look good

**Comment by dfuchss** [comment]
@Malyuk-A @willergeorg could you have a look at it again :)

**Review by willergeorg — APPROVED** [review]
Code changes since last review look reasonable to me.

**Review by MerlinMehmed — APPROVED** [review]
Tested on TS3. The testing steps worked fine. I have two suggestions that could be addressed in a follow-up PR.

The first of them is that I think it will be good for the user to have a confirmation message that the changes for the individual dates are saved after pressing the Save button. I can see that the button is disabled after the save is done but I think it might be confusing to some people.

Another thing I tried was to enter "Submissions" from the Exercises overview page as an instructor. When the regular due date has not passed yet I can see a warning message that the due date is not over, therefore I should wait before assessing. Then, when the regular due date is over but there is still one student for whom this date is not over, the message disappears. I can also see the submission of this student because this screen shows all the submissions.  
Probably, in this case, we can check the individual due dates and if there is at least one student who is still working we can show a message similar to "The due date has not passed for all students" and maybe suggest them in the message to use the Exercise Dashboard where we see only the submissions which have passed the due date

**Review by JohannesStoehr — DISMISSED** [review]
Tested on TS1 and especially focused on the programming exercises (for example the locking of the repositories). Everything worked as expected 👍

**Comment by b-fein** [comment]
@MerlinMehmed Thanks for the review!

> The first of them is that I think it will be good for the user to have a confirmation message that the changes for the individual dates are saved after pressing the Save button. I can see that the button is disabled after the save is done but I think it might be confusing to some people.

Good point! As that was an easy single-line addition, I added it in d40300f.

> Another thing I tried was to enter "Submissions" from the Exercises overview page as an instructor. When the regular due date has not passed yet I can see a warning message that the due date is not over, therefore I should wait before assessing. Then, when the regular due date is over but there is still one student for whom this date is not over, the message disappears. I can also see the submission of this student because this screen shows all the submissions. Probably, in this case, we can check the individual due dates and if there is at least one student who is still working we can show a message similar to "The due date has not passed for all students" and maybe suggest them in the message to use the Exercise Dashboard where we see only the submissions which have passed the due date

That one needs probably about another 100-200 LOC including tests scattered in various components (the submission page is different for each exercise type).
As the instructors should know the students who have got an individual due date (from experience *maximum* three per exercise), I think not showing a warning should be fine until I can add it in a follow-up.
The submissions are known and each knows its participation, therefore the necessary information should be available and the implementation should be possible without needing to fetch additional data from the server.

File: src/test/javascript/spec/component/course/course-exercises.component.spec.ts [hunk 4]
```diff
@@ -290,4 +296,71 @@ describe('CourseExercisesComponent', () => {
...
+        const participation = new StudentParticipation();
+        participation.individualDueDate = dayjs().add(3, 'days');
+        newExercise.studentParticipations = [participation];
+
+        component.course!.exercises![1] = newExercise;
+
+        component.activeFilters.clear();
+        component.toggleFilters([ExerciseFilter.OVERDUE]);
+
+        // ToDo: check that exercise is shown
```
Comment (MerlinMehmed) [inline src/test/javascript/spec/component/course/course-exercises.component.spec.ts:316]: Is this to do left for a later PR or you just forgot to remove it?

**Review by MerlinMehmed — APPROVED** [review]
Retested the displaying of the message. Works in both languages, Thank you for addressing the comment!
I also reviewed the code, looks good to me. I only left one small question

File: src/test/javascript/spec/component/course/course-exercises.component.spec.ts (hunk 4 above)
Comment (b-fein) [reply src/test/javascript/spec/component/course/course-exercises.component.spec.ts]: Good catch, I just forgot about it. Fixed in 9750ee4.

**Review by MerlinMehmed — APPROVED** [review]
Thank you for removing the TODOs. Reapprove!

**Review by JohannesStoehr — DISMISSED** [review]
Rechecked after the last commits for Merlins review.
Re-Approve

**Review by dfuchss — APPROVED** [review]
Changes since last review are looking good to me. Re-Approve

**Review by MerlinMehmed — APPROVED** [review]
Reapprove after merging the develop branch

**Review by LDAP — APPROVED** [review]
Tested on KIT-Test-Server. Code looks good.

**Review by jhainzlschmid — DISMISSED** [review]
I deployed this branch on TS3 and executed the Cypress E2E Tests against it. [All tests passed](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT-701). Therefore the most essential functionality still works 👍

**Review by JohannesStoehr — DISMISSED** [review]
Rechecked on TS1 and couldn't find any problems. Re-approve 👍

**Review by dfuchss — APPROVED** [review]
Change looks reasonable.

However client tests are still broken (but that's an issue on the `develop` branch)

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/ProgrammingExerciseScheduleService.java [hunk 5]
```diff
@@ -126,15 +134,16 @@ public void scheduleRunningExercisesOnStartup() {
...
+        else {
             // If a programming exercise got changed so that any scheduling becomes unnecessary, we need to cancel all scheduled tasks
             cancelAllScheduledTasks(exercise);
-            return;
         }
-        scheduleExercise(exercise);
     }
 
-    private static boolean needsToBeScheduled(ProgrammingExercise exercise) {
+    private boolean needsToBeScheduled(ProgrammingExercise exercise) {
```
Comment (Hialus) [inline src/main/java/de/tum/in/www1/artemis/service/scheduled/ProgrammingExerciseScheduleService.java:146]: Codacy also gives a warning here.

**Review by Hialus — CHANGES_REQUESTED** [review]
Reviewed all of the code.
Looks good to me, but there a two Codacy warnings. Please address them.

File: src/main/webapp/i18n/de/result.json [hunk 6]
```diff
@@ -71,7 +71,8 @@
                 }
             },
             "notLatestSubmission": "Hinweis: Dieses Ergebnis und Feedback bezieht sich <strong>nicht</strong> auf die letzte Einreichung.",
-            "seeMore": "Mehr anzeigen"
+            "seeMore": "Mehr anzeigen",
+            "afterDueDateFeedbackHidden": "Auch nach der Abgabefrist werden nicht immer Detailfeedbacks zu allen Testfällen angezeigt. Dies ist etwa der Fall wenn für andere Studierende eine spätere Abgabefrist gesetzt ist (z.B. zum Ausgleich von Krankheitstagen). Sobald alle Abgaben erfolgt sind, werden auch diese Feedbacks automatisch sichtbar."
```
Comment (ole-ve) [inline src/main/webapp/i18n/de/result.json:75]:
```suggestion
            "afterDueDateFeedbackHidden": "Auch nach der Abgabefrist werden nicht immer Detail-Feedbacks zu allen Testfällen angezeigt. Dies ist etwa der Fall, wenn für andere Studierende eine spätere Abgabefrist gesetzt ist (z.B. zum Ausgleich von Krankheitstagen). Sobald alle Abgaben erfolgt sind, werden auch diese Feedbacks automatisch sichtbar."
```

**Review by ole-ve — APPROVED** [review]
Reviewed code changes, look reasonable to me. Just a small suggestion

File: src/main/java/de/tum/in/www1/artemis/service/ResultService.java [hunk 7]
```diff
@@ -50,20 +46,22 @@
 
     private final AuthorizationCheckService authCheckService;
 
-    public ResultService(UserRepository userRepository, ResultRepository resultRepository, LtiService ltiService, ObjectMapper objectMapper, FeedbackRepository feedbackRepository,
+    private final ExerciseDateService exerciseDateService;
+
+    public ResultService(UserRepository userRepository, ResultRepository resultRepository, LtiService ltiService, FeedbackRepository feedbackRepository,
```
[the remark this answers is not in the document]
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/ResultService.java:51]: We have longer parameter lists in other classes due to constructor injection which is the preferred way in Spring Boot.
We should probably deactivate this warning in Codacy

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/ProgrammingExerciseScheduleService.java (hunk 5 above)
Comment (b-fein) [reply src/main/java/de/tum/in/www1/artemis/service/scheduled/ProgrammingExerciseScheduleService.java:146]: Addressed in the follow-up PR #4535.

File: src/main/webapp/i18n/de/result.json (hunk 6 above)
Comment (b-fein) [reply src/main/webapp/i18n/de/result.json:75]: Both ‘Detail’ and ‘Feedback’ are nouns present in the Duden dictionary. Therefore, I prefer the using a proper compound-noun without the hyphen.
